### PR TITLE
OCDS - Group validation errors by type

### PR DESCRIPTION
**Pull request by odscjames**
https://github.com/OpenDataServices/cove/issues/1159

### CODE REVIEW

**Comment by odscjames** [comment]
Here's a test file:

[ocds-data-not-grouped.json.zip](https://github.com/OpenDataServices/cove/files/3107086/ocds-data-not-grouped.json.zip)

Which produces:

cove-grouped

**Comment by odscjames** [comment]
(Closed by accident!)

Haven't run I18N process yet, want to get feedback from OCDS people first.

**Comment by odscjames** [comment]
(pr build fails, push build passes. This is due to having not done I18N strings yet - fine for now)

**Comment by odscjames** [comment]
http://1159-ocds-group-validation.dev.cove.opendataservices.coop/review/

File: cove_ocds/templates/cove_ocds/explore_base.html [hunk 1]
```diff
@@ -241,26 +241,79 @@ <h4 class="panel-title">
...
+      {% with error=error_json|json_decode %}
+        {% cove_modal_errors className="validation-errors-"|concat:error_prefix|concat:forloop.counter modalTitle=error.message errorList=values file_type=file_type full_table=True %}
+      {% endwith %}
+    {% endfor %}
+
+    <a name="validation-errors" class="anchor"></a>
+    <div class="panel panel-danger">
+      <div id="validation-panel" class="panel-heading pointer" role="region" aria-expanded="true" aria-controls="validationTable" data-toggle="collapse" data-target="#validationTable">
+        <h4 class="panel-title">
+          <span class="glyphicon glyphicon-collapse-up"></span>{% trans 'Structural Errors - Required' %}
```
Comment (duncandewhurst) [inline cove_ocds/templates/cove_ocds/explore_base.html]: 'Structural Errors - Required' -> 'Structural Errors - Required Fields'

File: cove_ocds/templates/cove_ocds/explore_base.html [hunk 2]
```diff
@@ -241,26 +241,79 @@ <h4 class="panel-title">
...
+        </h4>
+      </div>
+      <div id="validationTable" class="collapse in panel-body">
+        {% blocktrans %}Some or all of your entries were missing fields which are required by the OCDS Standard.{% endblocktrans %}
```
Comment (duncandewhurst) [inline cove_ocds/templates/cove_ocds/explore_base.html]: 'Some or all of your entries were missing fields which are required by the OCDS Standard.' -> 'Some or all of your data is missing fields which are required by the OCDS schema.'

File: cove_ocds/templates/cove_ocds/explore_base.html [hunk 3]
```diff
@@ -241,26 +241,79 @@ <h4 class="panel-title">
...
+
+    <a name="validation-errors" class="anchor"></a>
+    <div class="panel panel-danger">
+      <div id="validation-panel" class="panel-heading pointer" role="region" aria-expanded="true" aria-controls="validationTable" data-toggle="collapse" data-target="#validationTable">
+        <h4 class="panel-title">
+          <span class="glyphicon glyphicon-collapse-up"></span>{% trans 'Structural Errors - Format' %}
+        </h4>
+      </div>
+      <div id="validationTable" class="collapse in panel-body">
+        {% blocktrans %}Some or all of your entries have incorrectly formatted data.{% endblocktrans %}
```
Comment (duncandewhurst) [inline cove_ocds/templates/cove_ocds/explore_base.html]: 'Some or all of your entries have incorrectly formatted data.' -> 'Some or all of your data includes fields which are incorrectly formatted.'

**Comment by odscjames** [comment]
Changes made @duncandewhurst

File: cove_ocds/templates/cove_ocds/explore_base.html [hunk 4]
```diff
@@ -241,26 +241,79 @@ <h4 class="panel-title">
...
+
+    <a name="validation-errors" class="anchor"></a>
+    <div class="panel panel-danger">
+      <div id="validation-panel" class="panel-heading pointer" role="region" aria-expanded="true" aria-controls="validationTable" data-toggle="collapse" data-target="#validationTable">
+        <h4 class="panel-title">
+          <span class="glyphicon glyphicon-collapse-up"></span>{% trans 'Structural Errors - Other' %}
+        </h4>
+      </div>
+      <div id="validationTable" class="collapse in panel-body">
+        {% blocktrans %}Other validation errors.{% endblocktrans %}
```
Comment (duncandewhurst) [inline cove_ocds/templates/cove_ocds/explore_base.html]: 'Other validation errors.' -> 'Some or all of your data has validation errors.'

Comment (odscjames) [reply cove_ocds/templates/cove_ocds/explore_base.html]: Done

**Comment by odscjames** [comment]
Just to be totally clear @duncandewhurst where would you expect codelist errors to be?

**Comment by duncandewhurst** [comment]
@odscjames I think they should just appear in the codelist validation errors section but with the error count and location of first 3 errors which are currently missing from that section (see https://github.com/OpenDataServices/cove/issues/1044)

**Comment by odscjames** [comment]
(Notes for me: https://github.com/Julian/jsonschema/blob/master/jsonschema/validators.py#L473 here is a list of types, but we also have to include whatever we extend it with in .ve/src/libcove/libcove/lib/common.py and also note `validator_type = e.validator_value[0]` so we need to look up the range of values there! )

**Comment by duncandewhurst** [comment]
thanks @odscjames I checked and the errors in the 'required fields' and 'format' sections are correct, but there are still some errors under the 'other' heading which seem like they should be under 'format'

image

**Comment by duncandewhurst** [comment]
Also the currency codelist error is still duplicated under both 'structural errors - other' and 'codelist errors'

**Comment by odscjames** [comment]
@duncandewhurst Is it ok if we deal with the codelist issue in #1044 ? It will involve looking at different areas of code, was already a problem, and in the interests of keeping pull requests short and thus easier and thus more likely to actually be merged to master, that should probably be a separate pull request.

**Comment by duncandewhurst** [comment]
Yep no worries, I thought it might be an easy fix whilst looking at
grouping but no worries if not

**Comment by odscjames** [comment]
Thanks.

I have updated http://1159-ocds-group-validation.dev.cove.opendataservices.coop/review - do you want to have a look?

**Comment by duncandewhurst** [comment]
thanks @odscjames I think the only remaining inconsistency is that one "X is too short" error appears in under the "format" group whilst another "X is too short" error appears under the "other" group.

I think they should both appear under the "other" group:

image

image

**Comment by odscjames** [comment]
Ready for review

**Pull request merged**